Build: Assign `edit-post` global as `wp.editPost`

aduth opened the pull request:
Related: #4661

This pull request seeks to ensure module assignment into the `wp` global is camel-cased, specifically the `edit-post` module introduced in #4661 (as `wp.editPost`, currently `wp[ 'edit-post' ]`). As our build process is generalized, this required some customization of our Webpack build configuration, including the introduction of a custom Webpack plugin for enhancing tags to include custom keys (e.g. add `[dir]` in addition to `[name]`, `[id]`, etc). This plugin could be easily externalized to a separate module, likely under the WordPress namespace.

__Testing instructions:__

Verify that the build outputs file as expected and that the editor loads without errors.

aduth committed:
Build: Add Webpack plugin dirname tag transform

aduth committed:
Build: Assign edit-post library target as editPost

youknowriad reviewed (approved):
Not enough skills to review the webpack plugin properly but this seems to work great, I'm always impressed with the way you always find a solution to all issues :)

aduth committed:
Use basename on rawRequest for generating output directory

directory of resolved request is not sufficient because the resolved file of WordPress dependencies is within the `build-module` directory and would output as such. Instead format in such a way that we can call basename on the _raw_ requested path (e.g. `./editor` -> "editor", `./node_modules/@wordpress/hooks" -> "hooks")

gziolo commented on webpack.config.js:
```diff
@@ -75,10 +77,60 @@ const externals = {
+	 *
+	 * @param {Object} compiler Webpack compiler
+	 *
+	 * @return {void}
```
Do we really need to explicitly document that function doesn't return anything? Wouldn't it enough to assume that a function by default doesn't return value?

aduth replied:
>Do we really need to explicitly document that function doesn't return anything? Wouldn't it enough to assume that a function by default doesn't return value?

There's some precedent of this in existing core documentation. In general I'd be inclined to agree with you that it could be omitted for `undefined` return, but I could also see an argument for consistency / explicitness. Maybe worth bringing up in tomorrow's JS meeting? 😉

gziolo replied:
Docs again? Count me in ❤️

gziolo commented on webpack.config.js:
```diff
@@ -75,10 +77,60 @@ const externals = {
-			memo[ entryPointName ] = `./${ entryPointName }/index.js`;
+		entryPointNames.reduce( ( memo, entryPoint ) => {
+			entryPoint = castArray( entryPoint );
+			const [ name, path = name ] = entryPoint;
```
Just wanted to double check if I understand how destructuring works here:
* if there is a `name` and `path` provided they are set
* if the `path` is not set, we set its value as `name`

aduth replied:
>Just wanted to double check if I understand how destructuring works here:

This is correct. I've added a clarifying comment in d3a43bf.

gziolo replied:
Thanks, you don't always see defaults set here, using another param. Quite nice that is is possible 👍

gziolo commented on webpack.config.js:
```diff
@@ -75,10 +77,60 @@ const externals = {
+ *
+ * @see webpack.TemplatedPathPlugin
+ */
+class CustomTemplatedPathPlugin {
```
Do we plan to extract it and move to `WordPress/packages` at some point? It might be useful for other repositories.

aduth replied:
>Do we plan to extract it and move to `WordPress/packages` at some point? It might be useful for other repositories.

Yeah, seems like a generally useful plugin.

gziolo commented:
> /home/travis/build/WordPress/gutenberg/blocks/api/raw-handling/slack-markdown-variant-corrector.js
> [0]   9:4  error  Use @return instead  valid-jsdoc
> [0] 
> [0] ✖ 1 problem (1 error, 0 warnings)
> [0]   1 error, 0 warnings potentially fixable with the `--fix` option.

I see an error not related to this PR.

gziolo reviewed (approved):
I checked job logs on Travis. Everything looks as advertised. I can see that Webpack produces `editPost` bundle and Cypress' tests pass.

I don't have enough Webpack experience to fully validate plugin's code. However I'm satisfied with the result of this PR 👍

aduth committed:
Build: Clarify intent of entry definition destructuring
